### PR TITLE
Fix for DNS name resolution after performing init with --force-new-cluster

### PR DESCRIPTION
**Pull request by kylewuolle**
- What I did
This fixes a problem where the agent on a controller is stopped when a node leaves a swarm and is never restarted. I've added
a flag to the DaemonJoinsCluster method to indicate the case where a force init is being done. When that flag is set the existing agent is cleaned up
by setting the cluster provider to nil and waiting for the agent to stop. When the cluster provider is set after, the agent is setup properly. This PR 
fixes the following issue : https://github.com/docker/for-linux/issues/495

- How I did it
 Added a flag indicating that this is a force new cluster situation and the agent should be cleaned up before setting the cluster provider.
 
- How to verify it

1. Using the following Docker file build an image on each node called demo.

```
FROM ubuntu

RUN apt update
RUN apt install dnsutils -y

CMD /bin/bash -c "while true; do nslookup tasks.demo; sleep 2; done"
```

2. execute swarm init on one of the nodes
3. create a network `docker network create --scope swarm --driver overlay --attachable test`
4. create a service `docker service create --network test --mode global --name demo demo`
5. verify that the tasks.demo endpoint resolves to two ip addresses docker service logs demo
6. now execute docker swarm init --force-new-cluster on one of the nodes
7. demote and remove the other node and also, remove the service and network
8. recreate the service and network on the remaining node
9. have a third node join the remaining node
10. Previously, at this point node 3 would resolve tasks.demo to be it's container's ip but the tasks.demo would not resolve on the first node. Also the container on each node could not reach the container on the other node using it's ip. With this fix in place this would work as we expect and each task would be resolved to the respective ips.

- Description for the changelog
Fix a problem with DNS resolution after performing a cluster init with the --force-new-cluster option set

- A picture of a cute animal (not mandatory but encouraged)

### CODE REVIEW

File: daemon/daemon.go [hunk 1]
```diff
@@ -647,7 +647,15 @@ func (daemon *Daemon) registerLink(parent, child *container.Container, alias str
 
 // DaemonJoinsCluster informs the daemon has joined the cluster and provides
 // the handler to query the cluster component
-func (daemon *Daemon) DaemonJoinsCluster(clusterProvider cluster.Provider) {
+func (daemon *Daemon) DaemonJoinsCluster(clusterProvider cluster.Provider, forceNewCluster bool) {
+	// When forcing a new cluster, first clean up the existing agent
+	// ensuring that a new one will be created and started
+	if(forceNewCluster) {
+		daemon.setClusterProvider(nil)
```
Comment (fcrisciani) [inline daemon/daemon.go]: I still see this as not super clean. If the Agent is stopped, I believe libnetwork need to nullify the previous cluster provider automatically. Is there any other case where the clusterProvider is being reused without being set?

**Comment by thaJeztah** [comment]
Looks like there's a linting issue;

```
19:40:46 daemon/daemon.go:1::warning: file is not gofmted with -s (gofmt)
19:40:46 daemon/daemon.go:1::warning: file is not goimported (goimports)
19:40:47 Build step 'Execute shell' marked build as failure
```

File: daemon/daemon.go (hunk 1 above)
Comment (kylewuolle) [reply daemon/daemon.go]: You're right there's another way that might be better. I've updated https://github.com/docker/libnetwork/pull/2307. Now instead the cluster provider would be set to nil in agentClose.

**Comment by coolljt0725** [comment]
ping @kylewuolle  Jenkins failed

**Comment by thaJeztah** [comment]
@kylewoule should this one be closed now that https://github.com/docker/libnetwork/pull/2307 was merged (and will be vendored through https://github.com/moby/moby/pull/38983 ?)

Note that it was already included in Docker 18.09.4 https://github.com/docker/engine/pull/169

**Comment by thaJeztah** [comment]
oops, it was actually not yet included in 18.09; cherry-picking now

**Comment by thaJeztah** [comment]
ping @kylewuolle is this still needed now that docker/libnetwork#2307 was merged?
/cc @arkodg

**Comment by caoyj1991** [comment]
@thaJeztah @kylewuolle  Has the fixed merged into 18.09 docker ? and what can i do now?

**Comment by thaJeztah** [comment]
@caoyj1991 fix was backported in libnetwork through https://github.com/docker/libnetwork/pull/2354, and included in Docker 18.09.6 through https://github.com/docker/engine/pull/201 (should also be in Docker 19.03 and up)

**Pull request closed without merging**